Feature/domain managemnt multi tenant (#232)
MechanicalRock/mechanicalrock.github.io: _posts/2022-06-22-domain-management-for-multitenant-app-in-aws.markdown

> ---

Reviewer: A general comment (in case I miss in proofreading) - clients/users - it doesn't matter which you use, but ensure it's consistent across the article.

Multitenancy is a challenging problem as you have to consider your application's security, optimization, monitoring, and high availability. Each aspect of implementation could be a real challenge when it grows. For a part like domain management, you don't think it will be a problem as you will only require to do most work during the initial setup. 

However, it can be problematic along the way for a task such as SSL certificate renewal. If you have each certificate for each client's subdomain and each has a different expiration date, you need to manage them manually. Furthermore, for a task like setting up unique subdomains for new clients, what are the measures you take to ensure you will not impact your domains for other clients when there are changes to your current resources. When your number of clients grow, how do you have clear segregation of billing cost for each client? Lastly and most importantly, what is the disaster recovery plan for your instances, and how confident can you bring the application back online with minimal downtime? These are some overhead tasks we need to consider, and often most of the instances mentioned above are handled manually. 

> It might have more business value for you to focus on maintaining and developing the application, but managing the domain is critical in keeping the application live and accessible.

Reviewer 2: You list out a lot of potential challenges in the previous paragraph and then state here you are focusing on domain but what is the particular particular domain problem you are going to explain how to solve below?  Is it all of them?

Reviewer 2: I would also suggest making this introduction section shorter or putting a one sentence summary of the core issue at the start.  When I am reading posts I want to know in the first couple of lines if it addresses the problem I am trying to solve and if I therefore want to keep reading.

### Solution

> The diagram below is the solution to how you can typically start building it, but it will soon become challenging when it grows.

Reviewer 2: What is 'it'?

![Domain Management Architecture ](/img/blog/domain-management/architectureSingleAccount.png)

Generally, it is a good practice to segregate the production environment from the development environment, which should also apply to domain management. This approach has no segregation between environments and hosts all domain names within a single location. It increases the risk of impacting all tenants when there are service disruptions. 

The architecture diagram below shows how to manage domains for all your tenants in multiple environments.  

> ![Domain Management Architecture ](/img/blog/domain-management/architectureDiagram.png)

Reviewer 2: What are the key changes between these two diagrams?  Why have these changes been made?  How do they benefit us?  This should refer back to the core issue you have highlighted in your introduction and therefore how it is solved.

> AWS Route53 domain management provides the ability to segregate the management of domain names between accounts. DNS is a public-facing element you would want to minimize downtime as it will directly impact your revenue or reputation. Having segregation between accounts, especially for non-production and production, is key to reducing the risk of damage to production domains. It is also beneficial for segregating billing costs; hence, you will get better insights. 

Reviewer 2: This feels more like your 'Wrapping up' summary.

> AWS Route53 domain management provides the ability to segregate the management of domain names between accounts. DNS is a public-facing element you would want to minimize downtime as it will directly impact your revenue or reputation. Having segregation between accounts, especially for non-production and production, is key to reducing the risk of damage to production domains. It is also beneficial for segregating billing costs; hence, you will get better insights. 

Reviewer: > AWS Route53
This is the first time you have introduced AWS/Route53.  

* Is this article intended to be focussed on AWS?
* Could it apply for other services (and AWS is just your example)?

Depending on your intention for the above, be clear to identify the parts that are specific to an AWS implementation and which bits are more general.  For example, account management/structure is particular to AWS - you may manage it differently with other providers.

Not sure if you want to cover it in this article, but do you want to make an analogy with different hosting options / benefits - e.g. cloud services (like AWS) vs on-premise.  The way you look to manage domains may be different for each - creating new domains may be difficult in an on-prem environment.

Besides, you can configure a single CloudFront distribution to serve requests from multiple origins or subdomains and an SSL certificate to encrypt the information. A single distribution and SSL certificate can reduce the overhead for maintenance, such as renewing all certificates for each subdomain. 

#### Principle 

There are a few principles you should follow, which are listed below:

1. You should not provision resources under the top-level domains because your non-production and production account should be responsible for managing its subdomains and resources. 
2. Manage the registration of top-level domains from a central account. It is to ensure you have the main point to manage all domain names and subdomains should all be in its workload account to keep it consistent. 
3. All sub-domains for an application should live in the same workload account to reduce complexity. 

Lastly, a single domain management stack that manages all the resources makes change management more straightforward. You can then achieve the automation by setting up a CI/CD pipeline for the stack. 

#### Wrapping Up.

>  Setting up a robust pipeline for domain management is the first step in getting your multi-tenant cloud infrastructure right. 

Reviewer: The summary doesn't really match the content of the article for me.

* Is the intent of the post to demonstrate (with an example/tutorial) how to build a pipeline for domain management?
* Or is it to introduce a pattern for managing multi-tenant domains?
* Or something else?

Any of the above is a great topic, but don't try to do everything.

When I get to the end of an article, I'm looking for (depending upon the type of article):
* An 'ah-ha' moment from your article that will make me think/reflect.  
* A follow on question that will make me want to dive into the topic further.
* The answer to my immediate problem

---
layout: post
title: Domain Management for Multi-tenant Application in AWS
description: Architecting multi-tenant applications
date: 2022-07-12
author: Shermayne Lee
image: /img/blog/multitenancy/multitenant.png
tags: ['multitenant', 'route53', 'DNS', 'DomainManagement']
---

## Introduction

![Domain Management Multitenant](/img/blog/domain-management/multitenantDNS.png)

I have recently started building an application that has to manage multiple clients' domains within the same application instance. Having few clients doesn't cause any trouble with my implementation initially. When the number of clients grows, keeping track of the tasks needed to keep the domain stable is harder. Having clear segregation for each client is also challenging when a single application serves multiple clients. 

You might think it has more business value for you to focus on maintaining and developing the application, but managing the domain is critical in keeping it live and accessible. 

> Before we dive into the solution, what is a multitenant application? Multitenant application has a single instance of an application, and its infrastructure and business logic serves multiple clients. The book [Why Multi-Tenancy is Key to Successful and Sustainable Software-as-a-Service](https://books.apple.com/us/book/why-multi-tenancy-is-key-to-successful-and/id419723802) written by Larry Aiken explains the why it is essential to have multitenancy for some applications as compared to a single tenant.

Reviewer: ```suggestion
Before we dive into the solution, what is a multitenant application?  A multitenant application serves multiple distinct clients, or groups, using a single deployment of an application, its infrastructure and business logic. The book [Why Multi-Tenancy is Key to Successful and Sustainable Software-as-a-Service](https://books.apple.com/us/book/why-multi-tenancy-is-key-to-successful-and/id419723802) written by Larry Aiken explains the why it is essential to have multitenancy for some applications as compared to a single tenant.
```

Proof: flow suggestions.

It's interesting, and frustrating, that there appears to be no clear consensus on multitenancy, multi-tenancy or multi tenancy.

Going via [Lexico](https://www.lexico.com/search?utf8=%E2%9C%93&filter=en_dictionary&dictionary=en&query=multitenant), as per style guide, would suggest _multi-tenant_.

But this is a technology term: Wikipedia, Microsoft documentation and _shudder_ Gartner use _multitenancy_.

Not fussed which we use, so long as we're [consistent](https://mechanicalrock.slack.com/archives/C66SDG9S4/p1658214648902009)

Reviewer: - [ ] Use "multi-tenancy" throughout

> What makes multitenancy challenging, and so as domain management for multitenant? Multitenancy is a challenging problem because you must consider your application's security, optimization, monitoring, and high availability. Each aspect of implementation could be a real challenge when it grows.

Reviewer: ```suggestion
Why can multitenancy be challenging? Multitenancy is a challenging problem because you must consider your application's security, optimisation, monitoring, and high availability. As your client base grows, faults may have a higher severity.
```

Proofreading suggestion.

> In this post, I would love to share my thoughts on how you can design your application to increase the availability, visibility and reduce some repetitive tasks for each client within the same application instance for domain management.

Reviewer: Between these two, you haven't discussed why you may want to choose multitenancy.

Was this intentional?

If it was, consider making it explicit.:

> Discussion on whether multitenancy is right for you is outside the scope of this post.  Instead blah blah

### Solution

> It depends whether you are hosting your application on cloud or on premise. Cloud hosting has greater flexibility and scalability than on-premise hosting. Hence in this article, I will explain how you can implement your domain management system in AWS. 

Reviewer: ```suggestion
In this article, I shall focus on cloud hosting, in particular I will explain how you can implement your domain management system in AWS. 
```

Proofreading suggestion

> The diagram below is the solution to how you can typically start building it, but it will soon become challenging when the number of clients grows.

Reviewer: ```suggestion
The diagram below is the solution to one way you can start.  However, as the number of clients grows you will encounter issues.
```

Proofreading suggestion.

![Domain Management Architecture ](/img/blog/domain-management/architectureSingleAccount.png)

> Generally, it is a good practice to segregate the production environment from the development environment, which should also apply to your domain management system. This approach has no segregation between environments and hosts all domain names within a single location. It increases the risk of impacting all tenants when there are service disruptions. Hence it has a greater chance of affecting the availability of the application. 

Reviewer: ```suggestion
It is a good practice to segregate the production environment from the development environment.  This should also apply to your domain management system. The approach above has no segregation between environments: all domain names are hosted within a single location. It increases the risk of service disruptions: reducing availability and impacting all clients. 
```

Proofreading suggestion.

client/tenant?

The architecture diagram below shows how to manage domains for all tenants in multiple environments to minimize the likelihood of impacting your production environment.  

![Domain Management Architecture ](/img/blog/domain-management/architectureDiagram.png)

> The critical difference between these two diagrams is the latter has a clear separation between environments. The infrastructure for the non-production environment should be identical to the production environment to enable errors to be caught in the non-production environment first without impacting the production domains. DNS is a public-facing element you would want to minimize downtime as it will directly affect your revenue or reputation.

Reviewer: ```suggestion
There is now a clear separation between environments. The non-production environment should be identical to the production environment to enable errors to be caught in the non-production environment first without impacting the production domains. DNS is a public-facing element: you should aim to minimise downtime as it may directly affect your revenue or reputation.
```

Proofreading suggestions.

> AWS offers a highly scalable Domain Name System (DNS) Web service called AWS Route53. It provides the ability to segregate the management of domain names between accounts. Having segregation between accounts, especially for non-production and production, is key to reducing the risk of damage to production domains. It is also beneficial for segregating billing costs; hence, you will get better insights. 

Reviewer: ```suggestion
AWS offers a highly scalable Domain Name System (DNS) Web service called [AWS Route 53](https://aws.amazon.com/route53/). It provides the ability to segregate the management of domain names between accounts. Account segregation, especially between non-production and production, is key to reducing the risk of damage to production domains. It also provides better insight, by segregating billing costs. 
```

Proofreading suggestion.

Route53 or Route 53? AWS docs linked have a space. :shrug:

> Amazon CloudFront is a web service that speeds up the distribution of your static and dynamic web content. You can configure a single CloudFront distribution for each environment in AWS to serve requests from multiple origins or subdomains. It is essential to have SSL certificates configured for the subdomains to secure your web application. An SSL certificate is a digital certificate that authenticates a website's identity and enables an encrypted connection. You can have individual CloudFront distribution and SSL certificates for each subdomain, but having a single distribution and SSL certificate can reduce the overhead for maintenance, such as renewing only one SSL certificate for all subdomains.  

Reviewer: ```suggestion
[Amazon CloudFront](https://aws.amazon.com/cloudfront/), a content delivery network, speeds up the distribution of your static and dynamic web content. You can configure a single CloudFront distribution for each environment in AWS to serve requests from multiple origins or subdomains. It is essential to have [SSL certificates](https://en.wikipedia.org/wiki/Public_key_certificate) configured for each subdomains of your web application.  You can have separate CloudFront distributions, and SSL certificates, for each subdomain, but using a single distribution, and SSL certificate, can reduce the overhead for maintenance. For example, use a single SSL certificate across all subdomains.  
```

Proofreading suggestion.
If the reader has made it this far, you can assume they understand what SSL certificate is - but link out in case.
